clients/mongo/mongo.go

```diff
@@ -1,109 +1,189 @@
+		closingChannel  chan bool
+		initializeGroup sync.WaitGroup
+		pingOK          bool
+		clientMux       sync.Mutex
+		// pingOkMux       sync.Mutex
 	}
+)
 
-	var cs string
-	if config.Scheme != "" {
-	  cs = config.Scheme + "://"
-	} else {
-	  cs = "mongodb://"
-        }
+func NewStoreClient(config *Config, logger *log.Logger) (*StoreClient, error) {
+	if config.Timeout <= 0 {
+		return nil, errors.New("timeout is invalid")
+	}
+	mongoClient, err := newMongoClient(config)
+	if err != nil {
+		return nil, err
+	}
 
-	if config.User != "" {
-		cs += config.User
-		if config.Password != "" {
-			cs += ":"
-			cs += config.Password
-		}
-		cs += "@"
+	store := &StoreClient{
+		client:  mongoClient,
+		Context: context.Background(),
+		config:  config,
+		logger:  logger,
 	}
 
-	if config.Hosts != "" {
-		cs += config.Hosts
-		cs += "/"
-	} else {
-		cs += "localhost/"
+	return store, nil
+}
+func newMongoClient(config *Config) (*mongo.Client, error) {
+	connectionString, err := config.toConnectionString()
+	if err != nil {
+		return nil, err
 	}
+	clientOptions := options.Client().ApplyURI(connectionString)
+	mongoClient, err := mongo.Connect(context.Background(), clientOptions)
+	if err != nil {
+		return nil, err
+	}
+	return mongoClient, nil
+}
 
-	if config.Database != "" {
-		cs += config.Database
+// Mutex getters / setters
+func (s *StoreClient) getClient() *mongo.Client {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	return s.client
+}
+func (s *StoreClient) setClient(cli *mongo.Client) {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	s.client = cli
+}
+func (s *StoreClient) PingOK() bool {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	return s.pingOK
+}
+func (s *StoreClient) setPingOK(ping bool) {
+	s.clientMux.Lock()
+	defer s.clientMux.Unlock()
+	s.pingOK = ping
+}
+
+func (s *StoreClient) Start() {
+	if s.closingChannel == nil {
+		s.initializeGroup.Add(1)
+		go s.connectionRoutine()
 	}
+}
 
-	if config.Ssl {
-		cs += "?ssl=true"
-	} else {
-		cs += "?ssl=false"
+func (s *StoreClient) Close() error {
+	if s.closingChannel != nil {
+		s.closingChannel <- true
 	}
+	s.initializeGroup.Wait()
+	return s.getClient().Disconnect(s.Context)
+}
 
-	if config.OptParams != "" {
-		cs += "&"
-		cs += config.OptParams
+func (s *StoreClient) Ping() error {
+	if s.getClient() == nil {
+		return nil
 	}
-	return cs, nil
+	ctx, cancel := context.WithTimeout(context.Background(), s.config.Timeout)
+	defer cancel()
+	err := s.getClient().Ping(ctx, readpref.PrimaryPreferred())
+	s.setPingOK(err == nil)
+	return err
 }
 
-func Connect(config *Config) (*mgo.Session, error) {
-	connectionString, err := config.ToConnectionString()
+func (s *StoreClient) connectionRoutine() {
+	err := s.Ping()
+	var attempts int64
 	if err != nil {
-		return nil, err
-	}
-	dur := 20 * time.Second
-	if config.Timeout != nil {
-		dur = time.Duration(*config.Timeout)
+		s.logger.Printf("Unable to ping store initially : %v", err)
```

Comment:
trying to align comments with platform implementation
```suggestion
		s.logger.Printf("Unable to open inital store session : %v", err)
```